### ISSUE TITLE
write_tsdb to influxdb 1.2 opentsdb interface - collectd restart needed after upgrade from influxdb 1.1.1 to 1.2.0

### ISSUE THREAD

**Issue by martbhell** [collectd/collectd#2191]
*   Version of collectd: collectd-write_tsdb-5.7.1-1.el7.x86_64
*   Operating system / distribution: CentOS7

## Expected behavior
When influxdb is updated from 1.1.1 to 1.2.0 collectd will not have to be restarted.

## Actual behavior
It needs to be restarted

## Steps to reproduce

*   step 1 upgrade production influxdb to 1.2.0 from 1.1.1
*   step 2 observe working curl POST to the opentsdb interface and non-working collectd write_tsbd, no errors in normal log level collectd
*   step 3 restart collectd and see metrics again in grafana

## Question

I have however many servers with this kind of "stalled" collectd/write_tsdb.
Any hints for how to get more information out?

**Comment by rubenk** [comment]
compiling collectd with ./configure --enable-debug and setting the LogLevel to debug might show more details. A tcpdump of packets on the wire between collectd and influxdb might help too.

**Comment by martbhell** [comment]
tcpdump shows nothing when it's not working (it does if I run a curl to the opentsdb endpoint).
tcpdump on a machine where I've restarted collectd there's plenty of activity.

So collectd logs say write successful, but grafana shows no metrics.

Grepping away most of the tsdb logs:
<pre>
# grep tsdb /tmp/collectd_debug|grep -v interface|grep -v -e Writing -e "put"
[2017-02-24 13:36:22] plugin_load: plugin "write_tsdb" successfully loaded.
[2017-02-24 13:36:23] write_tsdb plugin: wt_flush_nolock: timeout = 0.000; send_buf_fill = 1393;
[2017-02-24 13:36:32] write_tsdb plugin: wt_flush_nolock: timeout = 0.000; send_buf_fill = 1366;
SNIP
[2017-02-24 13:38:22] write_tsdb plugin: wt_flush_nolock: timeout = 0.000; send_buf_fill = 1365;
[2017-02-24 13:38:22] write_tsdb plugin: error with wt_format_values
[2017-02-24 13:38:22] Filter subsystem: Built-in target `write': Dispatching value to the `write_tsdb/influxdb_obfu.example.com/4242' plugin failed with status -1.
[2017-02-24 13:38:22] Available write targets: ['aggregation', 'write_tsdb/influxdb_obfu.example.com/4242']
[2017-02-24 13:38:22] Filter subsystem: Built-in target `write': Plugin `write_tsdb/influxdb_obfu.example.com/4242' is back to normal operation. `write' succeeded.
[2017-02-24 13:38:22] write_tsdb plugin: wt_flush_nolock: timeout = 0.000; send_buf_fill = 1411;
[2017-02-24 13:38:22] write_tsdb plugin: wt_flush_nolock: timeout = 0.000; send_buf_fill = 1392;
SNIPSNIP
[2017-02-24 13:39:52] write_tsdb plugin: wt_flush_nolock: timeout = 0.000; send_buf_fill = 1411;
[2017-02-24 13:42:19] write_tsdb plugin: wt_flush_nolock: timeout = 0.000; send_buf_fill = 85;
[2017-02-24 13:42:19] write_tsdb plugin: wt_flush_nolock: timeout = 0.000; send_buf_fill = 0;
[2017-02-24 13:49:54] plugin_load: plugin "write_tsdb" successfully loaded.
[2017-02-24 13:49:55] write_tsdb plugin: wt_flush_nolock: timeout = 0.000; send_buf_fill = 1393;
SNIPSNIP
[2017-02-24 13:51:34] write_tsdb plugin: wt_flush_nolock: timeout = 0.000; send_buf_fill = 1351;</pre>

13:36:22 - started collectd with debug level
13:37:22 - got a metric in there
13:38:45 - yum downgraded influxdb to 1.1.0
13:38:59 influxdb 1.1.0 listening on :4242
 - not seeing any metrics
13:41 - yum upgrade to 1.2.0
13:41:09 - listening on port 4242
 - still not seeing any metrics
13:42:19 - killed the collectd process
13:49:54 - started collectd again
13:51:54 - see graphs in grafana, probably earlier

**Comment by rubenk** [comment]
> tcpdump shows nothing when it's not working 

I'd still like to see it. I'm interested in the behaviour when you shut down influxdb, if RST, FIN etc are send.

What is your interval set too?

> 13:38:45 - yum downgraded influxdb to 1.1.0

That doesn't explain the log 30 seconds earlier at
 [2017-02-24 13:38:22] Filter subsystem: Built-in target `write': Dispatching value to the `write_tsdb/influxdb_obfu.example.com/4242' plugin failed with status -1.

**Comment by martbhell** [comment]
I only ran tcpdump on the client and there were 0 packets on a stalled client, not during this debug client test. Interval 60s. I'll try again next week.

**Comment by rubenk** [comment]
Since collectd only sends packets every 60 seconds but influxdb restarts in 6 to 9 seconds on the same port, there might be a timing issue of some sorts. Influxdb should send FINs when it goes down, and the new influxdb should send a RST when it gets packets from a previous session. So a full tcpdump from both sides and full debug log would be most helpful.

**Comment by martbhell** [comment]
Okay, I have a timeline as below, full tcpdumps and collectd debug logs.
I created tcpdump filtered pcap files with these filters in wireshark "!arp and !vrrp and tcp.dstport != 5353 and !ssh"
Are you happy with the timeline or should I try something more?
A yum downgrade/upgrade does not restart influxdb.
I can put the files in a link, should I e-mail this to the e-mail in your github profile @rubenk ?

09:03:13 - yum downgrade to 1.1.1 influxdb
[opentsdb] 2017/02/27 09:03:52 Listening on: [::]:424
09:04:45 : tcpdumps started with:
 tcpdump -i eth0 -w /tmp/collectd.influxdb_host.6.pcap and tcpdump -i em1 -w /tmp/collectd.client.6.pcap
io3.6
09:05:07 - stopped client collectd
[2017-02-27 09:05:51] type = logfile, key = File, value = /tmp/collectd_debug.6
09:06:10 - see metrics in grafana (interval is 10s)
09:06:50~ - yum update influxdb to 1.2.0
09:06:51: last metric in grafana
09:06:55: systemctl restart influxdb
2017-02-27T07:07:35Z Listening on: [::]:4242 service=opentsdb
09:08:16: still no metric in grafana
09:08:50: still no metric in grafana
2017-02-27 09:09:12: sent grafana "annotation" to opentsdb, showed up instantly
09:08:33~  : ctrl +C collectd client
[2017-02-27 09:09:51] type = logfile, key = File, value = /tmp/collectd_debug.6
09:09:51: metric visible in grafana
09:10~ yum downgrade influxdb
09:10:40~ systemctl restart influxdb
2017-02-27T07:10:39Z openTSDB TCP listener closed service=opentsdb
[opentsdb] 2017/02/27 09:11:17 Listening on: [::]:4242
09:12:55: still no metrics in grafana
2017-02-27 09:13:05: sent grafana "annotation" to opentsdb, showed up instantly
[2017-02-27 09:14:04] type = logfile, key = File, value = /tmp/collectd_debug.6
and metric showed up at 09:14:04 in grafana

**Comment by rubenk** [comment]
yeah, mail is fine, but if you could put the pcaps on https://www.cloudshark.org/ I'd prefer that.

I'm having a bit of trouble making sense of your timeline. Did you hand-edit it?
For example, first I see `[opentsdb] 2017/02/27 09:03:52 Listening on: [::]:424` but later after you upgrade influxdb `[opentsdb] 2017/02/27 09:11:17 Listening on: [::]:4242`

And further:
```
09:05:07 - stopped client collectd
[2017-02-27 09:05:51] type = logfile, key = File, value = /tmp/collectd_debug.6
09:06:10 - see metrics in grafana (interval is 10s)
```
if you stopped collectd, how can you receive metrics in grafana?

So, let's try a singe action first
- start tcpdump, capturing traffic from / to the influxdb port
- start collectd
- upgrade influxdb
- stop tcpdump

**Comment by martbhell** [comment]
Yes the whole timeline is hand edited. I'll normalize the timestamps.

Cloudshark free trial is too small for this, compressed the logs are 31MB.
The servers are in production but at least the collectd client logs are small. 
I'll send you a link over e-mail with the logs.
I'll also update the timeline. This:

<pre>
[2017-02-27 09:05:51] type = logfile, key = File, value = /tmp/collectd_debug.6
</pre>

Is that "./sbin/collectd -f" outputs.

And this is from "journalctl -xeu influxdb" which I think it outputs when the opentsdb receiver is listening.
<pre>
[opentsdb] 2017/02/27 09:11:17 Listening on: [::]:4242
</pre>
